server/pbenchinacan/run-pbench-in-a-can

```diff
@@ -39,6 +39,41 @@ TMP_DIR=${TMP_DIR:-${WORKSPACE_TMP:-/var/tmp/pbench}}
 export SRV_PBENCH=${SRV_PBENCH:-${TMP_DIR}/srv/pbench}
 export PB_DEPLOY_FILES=${PB_DEPLOY_FILES:-${TMP_DIR}/pbench_server_deployment}
 
+# This function removes host files/directories which are owned by the pbench
+# user inside the can and therefore may not be accessible outside the can.
+function remove_from_can {
+    # First, try to remove the target(s) from outside the container, but
+    # suppress all errors.  If (any of) the target(s) still exist(s), try
+    # removing it/them from inside the container; display the results, but
+    # suppress error handling.  Finally, try again to remove the target(s) from
+    # outside the container:  the result of this operation will be the return
+    # status of the function (if the target(s) was/were successfully removed
+    # inside the container, then this will result in a success status from this
+    # function; otherwise, it will trigger an error if it is unsuccessful).
+    local outside_dir=${1}; shift
+    local inside_dir=${1}; shift
+
+    pushd ${outside_dir} 1>/dev/null
+    rm -rf "${@}" 1>/dev/null 2>&1 || true
+    for d in "${@}"; do
+        if [[ -e ${d} ]]; then
+            podman run \
+                --entrypoint rm \
+                --pull ${PB_SERVER_IMAGE_PULL_POLICY} \
+                --rm \
+                --volume ${outside_dir}:${inside_dir}:Z \
+                --workdir ${inside_dir}
+                ${PB_SERVER_IMAGE} \
+                -rf "${@}" \
+                || true
+            break  # The first call _should_ get them all; continuing is pointless.
+        fi
+    done
+    rm -rf "${@}" && rv=0 || rv=1
```

Comment:
Under what circumstances would this succeed if the identical statement at line 57 failed *and* the in-container `rm` also failed??

Also, the `shift`, `shift`, `${@}` suggests that documenting the function's args in its "docstring" comment might not be a bad idea ...

Reply:
> Under what circumstances would this succeed if the identical statement at line 57 failed and the in-container rm also failed??

If the target is a directory on the host which contains one or more files or directories created inside the container, then the initial attempt to delete the target will fail because the directory is not empty; then the attempt to delete the target inside the container should manage to remove the files/directories previously created inside the container, but it might fail to remove the target itself if, for instance, the user inside the container lacks sufficient access to it; however, now that the files/directories created inside the container have been removed, the final attempt from outside the container should succeed.

Nevertheless, the real purpose of the "third attempt" is to create the appropriate exit status for the function.  So, if either of the first two attempts fails (and the situation is not reconciled by the subsequent attempts), the final `rm` will create the appropriate `rv` value.  (I can't simply use `${?}`, because we're running under `bash -e`, and we won't get the chance to return from the function if the `rm` fails, and I want to make sure that the `popd` gets run.)

> Also, the shift, shift, ${@} suggests that documenting the function's args in its "docstring" comment might not be a bad idea ...

Yes....